### PR TITLE
Adversarial workloads: event size, dynamic fields

### PR DESCRIPTION
**Pull request by oxade**
## Description 

Add MaxEevents and MaxDynamicField workloads.

While this pr is open, I'll keep adding more workloads; or they'll come in subsequent prs.
Still trying to push some of the numbers higher as my machine locks up when I try, but wanted to open up the PR for folks to also play with it.

Working on MaxCommands, and a few others

## Test Plan 

N/A
---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-benchmark/src/workloads/data/adversarial/sources/adversarial.move
Comment (sblackshear) [inline crates/sui-benchmark/src/workloads/data/adversarial/sources/adversarial.move:109]:
```suggestion
    public fun emit_events(n: u64, size: u64) {
```

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (sblackshear) [inline crates/sui-benchmark/src/workloads/adversarial.rs:268]: For this test, what I had in mind was not adding dynamic fields (which basically looks the same as object creation from the runtime's perspective), but reading a bunch of dynamic fields added by a previous tx (which is a good stress test because it's the only time the Move VM needs to go to the DB).

For this, I'd suggest:
- adding a module initializer to `adversarial` that creates one object, then adds a large number of dynamic fields to it
- adding this object ID as a field in`TestPayload`
- make this logic call a function that reads N dynamic fields from that object

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (sblackshear) [inline crates/sui-benchmark/src/workloads/adversarial.rs:280]: I would say: let's emit events that are as big as possible (do we have an event size limit?). More generally, whenever we have a choice In these adversarial payloads, I would suggest going with the choice that is hardest for the system to handle (bigger objects, more objects, bigger events, ...) without bumping into a limit somewhere.

**Review by sblackshear — COMMENTED** [review]
Awesome, love seeing these new payloads!

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (sblackshear) [reply crates/sui-benchmark/src/workloads/adversarial.rs:268]: Besides dynamic field reads, I think another good on would be trying to read a lot of shared objects at once. For that I think the recipe is similar: create the shared objects in a module initializer, remember them in the payload, use them in the logic

File: crates/sui-benchmark/src/workloads/data/adversarial/sources/adversarial.move
Comment (oxade) [reply crates/sui-benchmark/src/workloads/data/adversarial/sources/adversarial.move:87]: The problem here is that event size a sum of the type tag size and the object size. Type tag and hence it's size are not exposed in Move, so either we need to create a new native to expose type tag size, or event size or something like that.
In the meantime though, I manually verified the sizes and have [snapshot](https://github.com/MystenLabs/sui/pull/9493/files#diff-6d585c35e0871d8704ee5b0f0e0b3f8ee286e197cbaa8b0a70b5f7a8d8748372) tests to detect when something changes.

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (oxade) [reply crates/sui-benchmark/src/workloads/adversarial.rs:280]: We quickly run out of gas when emitting events that are both large and plentiful, making us unable to hit either limit hence why I separated it into two tests: one that emits max number of small events, and one that emits a low number of max-sized events.

Comment (oxade) [reply crates/sui-benchmark/src/workloads/adversarial.rs]: Condensed both now

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (oxade) [reply crates/sui-benchmark/src/workloads/adversarial.rs:304]: Good suggestions. I did the first but was only able to add 33 dynamic fields before running out of gas. Will improve this in a subsequent PR.
Will also add the shared object read suggestion. In the meantime, I want to unblock this PR.

File: crates/sui-benchmark/src/workloads/data/adversarial/sources/adversarial.move
Comment (sblackshear) [inline crates/sui-benchmark/src/workloads/data/adversarial/sources/adversarial.move:97]: Why are these casts needed (isn't `i` already a `u64`)?

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (sblackshear) [inline crates/sui-benchmark/src/workloads/adversarial.rs:74]: Let's clean this up before landing? This printing can be added to the "Adversarial transactions should never abort" message if it's helpful.

File: crates/sui-benchmark/src/workloads/data/adversarial-dynamic-fields/sources/adversarial_dynamic_fields.move
Comment (sblackshear) [inline crates/sui-benchmark/src/workloads/data/adversarial-dynamic-fields/sources/adversarial_dynamic_fields.move:1]: Could you put this as an additional module in the `adversarial` package rather than creating an extra package? That will save us from having to do an extra publish command, remembering an extra package ID, etc.

Although: I don't actually see the extra publish command, so unsure how this is working right now...

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (sblackshear) [inline crates/sui-benchmark/src/workloads/adversarial.rs:242]:
```suggestion
                    self.df_parent_obj_ref = o.0; 
                    break
```

**Review by sblackshear — CHANGES_REQUESTED** [review]
Requesting changes for the dynamic fields workload--I can't see how it's working given that we're not publishing the package.

**Review by sblackshear — COMMENTED** [review]
Ah, I see--I think you just want to dele `data/adversarial-dynamic-fields` and then we're good.

File: crates/sui-benchmark/src/workloads/data/adversarial-dynamic-fields/sources/adversarial_dynamic_fields.move
Comment (oxade) [reply crates/sui-benchmark/src/workloads/data/adversarial-dynamic-fields/sources/adversarial_dynamic_fields.move:1]: Ah you can ignore this file. It's WIP and made it in accidentally.
The reason for this new package was that we were running out of gas in the module initializer of the original package, so we could not create enough dynamic fields. My hope was that if we publish a smaller package just for dynamic field testing, we can have more gas left for the initializer, but this is WIP.

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (sblackshear) [inline crates/sui-benchmark/src/workloads/adversarial.rs:98]: This is a bit funky--I would suggest having `get_payload_args` take `self` and adding the shared object arg inside there instead. This will also work better for future workloads with other special args.

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (sblackshear) [reply crates/sui-benchmark/src/workloads/adversarial.rs:304]: Rather than publishing a separate package, I would suggest calling `add_n_dynamic_fields` one time, inside `Workload::init`. That will maximize the gas you have available for initialization, and will be simpler than adding a separate package.

**Review by sblackshear — APPROVED** [review]

File: crates/sui-benchmark/src/workloads/adversarial.rs
Comment (sblackshear) [inline crates/sui-benchmark/src/workloads/adversarial.rs:209]: Nit: you could get away with keeping an `(ObjectID, SequenceNumber)` pair here.

File: crates/sui-benchmark/tests/simtest.rs
Comment (oxade) [inline crates/sui-benchmark/tests/simtest.rs:238]: @mystenmark this line seems to be causing failures in CI. Any ideas how we can get around it?

**Pull request merged**
